BugFix: fix encoding problems with QTextStreams

SlySven opened the pull request:
As Leris/Kebap found in issue #5262 Windows in particular does not default to using UTF-8 when reading and writing text files - instead it can use an 8-bit local which (probably) depends on the end-users OS country settings. However this is not UTF-8 and that will cause issues if such files are used on another system or OS that does not use the same setting.

This commit forces all uses of `QTextStream` (that are not associated with a `QString` as the underlying "device" for which it is not applicable) to use the UTF-8 encoding.

However retro-fitting this change *may* (checks will be useful) mess with existing files, specifically:
* The config.lua file that contains details of packages/modules in the `.mpackage` ZIP archive file format. This is the bug that this is intended to cure!
* Reading of the external Lua files that Mudlet uses itself, the `LuaGlobal.lua` file and the others that it itself loads - however since those are pure ASCII there should not be an issue with them.
* (WIndows ONLY) Reading the external `utf8_filenames.lua` file that is used to patch the Lua IO handlers to work with the UTF-16 file names that Windows uses in non-en locales. There shouldn't be an issue with this.
* Custom user dictionaries - it is possible that the existing code was making use of UTF-8 anyhow - it needs someone like Leris/Kebap to check that any custom dictionaries still work given that they had a system this encoding problem was happening...!

Provided it doesn't break other things this should close #5262.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
BugFix: fix encoding problems with QTextStreams

As Leris/Kebap found in issue #5262 Windows in particular does not default
to using UTF-8 when reading and writing text files - instead it can use an
8-bit local which (probably) depends on the end-users OS country settings.
However this is not UTF-8 and that will cause issues if such files are
used on another system or OS that does not use the same setting.

This commit forces all uses of `QTextStream` (that are not associated with
a `QString` as the underlying "device" for which it is not applicable) to
use the UTF-8 encoding.

However retro-fitting this change *may* (checks will be useful) mess with
existing files, specifically:
* The config.lua file that contains details of packages/modules in the
  `.mpackage` ZIP archive file format. This is the bug that this is
  intended to cure!
* Reading of the external Lua files that Mudlet uses itself, the
  `LuaGlobal.lua` file and the others that it itself loads - however since
  those are pure ASCII there should not be an issue with them.
* (WIndows ONLY) Reading the external `utf8_filenames.lua` file that is
  used to patch the Lua IO handlers to work with the UTF-16 file names that
  Windows uses in non-en locales. There shouldn't be an issue with this.
* Custom user dictionaries - it is possible that the existing code was
  making use of UTF-8 anyhow - it needs someone like Leris/Kebap to check
  that any custom dictionaries still work given that they had a system
  this encoding problem was happening...!

Provided it doesn't break other things this should close #5262.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
BugFix: correct a couple of typos!

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

Kebap commented:
When I try import the same package file in this version now, Mudlet still shows the diamonds, whereas Sublime Text editor shows the correct special characters, as seen in screenshot above.

Did not try to export a new package with this version, yet.

Kebap commented:
> Did not try to export a new package with this version, yet.

Created a new package in Windows under 5264 dev version, included some umlauts like äöü
Imported said package in Windows under 5264 dev version, display is fine as expected.
Imported same package in Linux release version, still seeing those diamonds.

Am I doing it wrong? 
Does this fix expect the exporter or the importer to use the latest dev version or both?

SlySven commented:
> > Did not try to export a new package with this version, yet.
>
> Created a new package in Windows under 5264 dev version, included some umlauts like äöü
Imported said package in Windows under 5264 dev version, display is fine as expected.
Imported same package in Linux release version, still seeing those diamonds.

:thinking: Humm, I'd have hoped that the Linux release version would be using UTF-8 - but perhaps not, in which case it is probably using ***Latin1*** (a.k.a. "ISO 8859-1") - that being the case you'd need to check that a Linux build of this PR.

>
> Am I doing it wrong?
> Does this fix expect the exporter or the importer to use the latest dev version or both?

Not sure what you mean by "latest dev version" - it "fixes" *both* exporter AND importer and makes them both use UTF-8. I suppose it would be worth checking precisely what encodings have been used in your "faulty" packages - can you post up a sample file {ideally with the same content and with a clear indication of what the sample messed up character are supposed to be} from both your Windows AND Linux **release** versions and I will try to determine what encodings were/are used on your systems for both of them - it seems that A) they are NOT the same and B) *neither* is UTF-8 - how we get around that for *existing* packages is going to take a bit more thought!

Kebap commented:
> can you post up a sample file

Eventually, starting here is a package created in this very latest dev version on Windows, how it's supposed to look (with umlauts ä, ö, ü, ÄÖÜ) and the resulting package files:

![image](https://user-images.githubusercontent.com/117238/121641797-eff6e000-ca8f-11eb-87fe-18f622b77f37.png)

[äöü.xml.txt](https://github.com/Mudlet/Mudlet/files/6636311/aou.xml.txt)
[config.lua.txt](https://github.com/Mudlet/Mudlet/files/6636312/config.lua.txt)

Seems fine on Windows still:
![image](https://user-images.githubusercontent.com/117238/121641974-292f5000-ca90-11eb-82d3-3a005348b1ff.png)

Kebap commented:
> Not sure what you mean by "latest dev version"

From this PR or PTB or release version

> how we get around that for _existing_ packages is going to take a bit more thought!

Exactly. How do we fix affected packages created since release of 4.12 and now every day while we struggle to fix. 

Worst case: existing packages can't be fixed. Creator needs to update and create a new package. Player needs to update Mudlet and download new package first.

SlySven commented:
in the `config.lua` file the author entry is a **single** `0xe4` byte - which is the `ä` character in the Latin1 (and a few other) encodings. It is NOT a UTF-8 encoding. 

That is not what should be there if it was generated with the source code of **this** PR - but **is** what I would expect from the current `development` branch code... so your "created in this very latest dev version on Windows," confuses me.

I would be interested to see what your `development` branch build on Linux (or indeed the release version) makes if run to make the same files as you have here on Windows.

Kebap commented:
> That is not what should be there if it was generated with the source code of **this** PR - but **is** what I would expect from the current `development` branch code... so your "created in this very latest dev version on Windows," confuses me.

How do we continue then? 

The top left Mudlet icon is red, that is proof enough, no?

SlySven commented:
> The top left Mudlet icon is red, that is proof enough, no?

It will be the red "development" version icon for *both* builds of the main `development` branch code AND a test build of this PR!

Also - unless one predefines the `MUDLET_VERSION_BUILD` environmental variable (and ensures that all the object files that make use of it - and some changes over the last couple of years has made that list much larger than it used to be!) then both builds will also end up with a plain `-dev` tagged onto the end of the version. If you are using a CI built one - i.e. the files listed in the second post here ("Test versions") then they do get a more informative identifier; however a self-build on one's own system will not by default!

Kebap commented:
SlySven, I am telling you, I was using the provided download in this PR by the name of "Mudlet-4.11.3-testing-pr5264-1deda93f-windows.zip" and did not compile any other development version just for shits and giggles. I implore you to reconsider.

Kebap commented:
Here, I did it again:

![image](https://user-images.githubusercontent.com/117238/121774264-f6658480-cb81-11eb-849d-4dfbe0037621.png)

[config.lua.txt](https://github.com/Mudlet/Mudlet/files/6642070/config.lua.txt)
[öüä.xml.txt](https://github.com/Mudlet/Mudlet/files/6642071/oua.xml.txt)

SlySven commented:
So... the `config.lua` file seems to have been encoded in, (probably) Latin1/ISO 8859-1 - though you had to add a `.txt` extension to get it uploaded to here. That does seem to be wrong so I am going to have to check my code... *sigh*

Kebap commented:
> you had to add a `.txt` extension to get it uploaded to here

Does that create any problems? Would you like to receive the package elsewhere as well?
 I obviously didn't do that while noticing the original problem. 
Will try and provide the Linux equivalents for comparison asap.

SlySven commented:
😕 This is weird - I've just run this PR's code on a local (Mingw-w64/MSYS2) build on one of my Windows 10 OSes. The config.lua file is: (with a `.txt` extension added on so that the GH system will allow it to be uploaded here) [config.lua.txt](https://github.com/Mudlet/Mudlet/files/6651392/config.lua.txt).

As expected however - it DOES use a UTF-8 encoding- note the multiple uses of the `ä` for the package name and author:
![image](https://user-images.githubusercontent.com/6163092/121961944-f7540d00-cd5f-11eb-9507-582b26f02c68.png)

That `ä` character is still  the U+00E4 {LATIN SMALL LETTER A WITH DIAERESIS} which encodes to the two bytes 0xC3 0xA4 in UTF-8 (and of course 0xE4 in Latin1/ISO 8859-1 - and some other encodings)!

I will just run it again to see what it looks like when re-opened in the package exporter...

Kebap commented:
I can confirm, your ä looks like UTF to me as well and therefore not like the ä that I produced before.

SlySven commented:
So what could cause the difference? I will try the above test build (the Windows zip file in the "Add deployment links") one in a bit to see what *that* produces...

Kebap commented:
It seems like your adjustments didn't actually change behavior at all, but I didn't verify sources, let alone know how those functions work. Can you confirm that UTF is really used in any case for writing the package file? Because it seems to be not.

SlySven commented:
Ah,  how many times did you try this and did you remove any previous temporary or `.mpackage` files between runs? Although we call it a *Package Exporter* ~~I think you have to use the "Module Manager" if you want to have it available in the former for updating~~{*Edit: no, perhaps not!*} - and perhaps the updating part is not working (it didn't in the dim and distant past because of a code defect IIRC)...

Curiously if the `.mpackage` file is loaded with the module or the package manager it becomes available as an existing item in the package exporter - but the details (of the author at least, didn't check anything else) only seem to be reloaded if it is loaded via the **package** manager.

> ... Can you confirm that UTF is really used in any case for writing the package file? ...

Yes I blooming well can!

When writing the `config.lua` file out to disk - see the change here:
![image](https://user-images.githubusercontent.com/6163092/121967880-130fe100-cd69-11eb-8871-c6a9c06122dc.png)

and this ***is*** what is needed - see https://doc.qt.io/qt-5/qtextstream.html#details , specifically:
![image](https://user-images.githubusercontent.com/6163092/121968291-f922ce00-cd69-11eb-981d-1aa2fbcbcd85.png)

SlySven commented:
And, for the record, I was also able to create a (differently named) package/module using a different character in the package name and author with the Windows test build available above:
![image](https://user-images.githubusercontent.com/6163092/121969255-c8439880-cd6b-11eb-9541-25920006f6c8.png)

Kebap commented:
I did not repeat often, nor update existing packages. I created them anew. In fact, I deliberately put new file names each time to not confuse experiments.

Here, I made a video for you to see how and where exactly I click to receive this result in the very Windows test build available above:

![export_utf](https://user-images.githubusercontent.com/117238/121970129-24131f00-cd76-11eb-9836-b1e00c3964a5.gif)

SlySven committed:
Testing: report of name of QTextCodec actually used to write config.lua

This is to test what is used on @Kebap (Leris) system which doesn't seem to
work with the "fixes" this PR is supposed to introduce...

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven commented:
Trying to follow what is happening (I had to find something that can playback animated GIFs in a controllable manner and resorted to **IrfanView** https://www.irfanview.com/ - which uses `G` to stop/resume playback) - I see that you open the `config.lua` file that you extracted from the manufactured package **first of all** in Windows `Notepad.exe` and that identifies the file's contents as "ANSI" (yet my copy, produced in the same manner, **is** UTF-8 😖) :
![image](https://user-images.githubusercontent.com/6163092/122651119-9f861f00-d12e-11eb-9dc8-ba9ea4e14e8f.png)
then you tried the same file in, I believe, the "Sublime" text editor - which confirmed that the encoding used for your file was indeed NOT UTF-8...

I have added something that will report the codec that got used to write the `config.lua` file - please @Kebap check the latest build and check that it reports "UTF-8" (and gives a nice long list of alternatives...) - i.e. check the things with the green markngs:
![image](https://user-images.githubusercontent.com/6163092/122656948-12a38b80-d157-11eb-9bf9-8b3a8e5a0d1f.png)

Also, notice how we use the package name directly as the file-name (red markings) - that strike me as dangerous given that the end-user could use characters in the name that are bad/invalid/unwise in **filenames** (e.g. `/` or `:` or ` ` or `;` etc. etc.)

Kebap commented:
Here is the red text for you:

```
System Message: Using QTextCodec: "System" from the available list:
"437",
"667",
"737",
"869",
"ASMO-708",
"Apple Roman",
"Big5",
"Big5-ETen",
"Big5-HKSCS",
"CP1250",
"CP1251",
"CP1252",
"CP1253",
"CP1254",
"CP1255",
"CP1256",
"CP1257",
"CP1258",
"CP819",
"CP850",
"CP866",
"CP874",
"CP936",
"CP949",
"CP950",
"ECMA-114",
"ECMA-118",
"EUC-JP",
"EUC-KR",
"GB18030",
"GB2312",
"GBK",
"IBM437",
"IBM819",
"IBM850",
"IBM866",
"IBM869",
"IBM874",
"ISO 8859-11",
"ISO 8859-8-I",
"ISO-2022-JP",
"ISO-8859-1",
"ISO-8859-10",
"ISO-8859-10:1992",
"ISO-8859-13",
"ISO-8859-14",
"ISO-8859-15",
"ISO-8859-16",
"ISO-8859-2",
"ISO-8859-3",
"ISO-8859-4",
"ISO-8859-5",
"ISO-8859-6",
"ISO-8859-6-I",
"ISO-8859-7",
"ISO-8859-8",
"ISO-8859-9",
"JIS7",
"KOI8-R",
"KOI8-RU",
"KOI8-U",
"MS936",
"MS_Kanji",
"M_CP437",
"M_CP667",
"M_CP737",
"M_CP869",
"MacRoman",
"Mazovia",
"SJIS",
"Shift_JIS",
"System",
"TIS-620",
"TSCII",
"UTF-16",
"UTF-16BE",
"UTF-16LE",
"UTF-32",
"UTF-32BE",
"UTF-32LE",
"UTF-8",
"WINSAMI2",
"WS2",
"arabic",
"cp-gr",
"cp437",
"cp6677",
"cp737",
"cp869",
"csHPRoman8",
"csIBM866",
"csIBM869",
"csISOLatin1",
"csISOLatin2",
"csISOLatin3",
"csISOLatin4",
"csISOLatin5",
"csISOLatin6",
"csISOLatinArabic",
"csISOLatinCyrillic",
"csISOLatinGreek",
"csISOLatinHebrew",
"csKOI8R",
"csPC850Multilingual",
"csPC8CodePage437",
"cyrillic",
"greek",
"hebrew",
"hp-roman8",
"ibm-869",
"ibm-869_P100-1995",
"iscii-bng",
"iscii-dev",
"iscii-gjr",
"iscii-knd",
"iscii-mlm",
"iscii-ori",
"iscii-pnj",
"iscii-tlg",
"iscii-tml",
"iso-celtic",
"iso-ir-100",
"iso-ir-101",
"iso-ir-109",
"iso-ir-110",
"iso-ir-126",
"iso-ir-127",
"iso-ir-138",
"iso-ir-144",
"iso-ir-148",
"iso-ir-157",
"iso-ir-199",
"iso-ir-226",
"latin1",
"latin10",
"latin2",
"latin3",
"latin4",
"latin5",
"latin6",
"latin8",
"latin9",
"macintosh",
"roman8",
"windows-1250",
"windows-1251",
"windows-1252",
"windows-1253",
"windows-1254",
"windows-1255",
"windows-1256",
"windows-1257",
"windows-1258",
"windows-869",
"windows-936",
"windows-949".
```

> Also, notice how we use the package name directly as the file-name (red markings) - that strike me as dangerous given that the end-user could use characters in the name that are bad/invalid/unwise in **filenames** (e.g. `/` or `:` or ` ` or `;` etc. etc.)

You may open a new issue to discuss this further, not to inflate this already complex one, but using those character will invoke a "file could not be opened for writing" error message and not crash Mudlet at all.

SlySven commented:
Oh *bugger*! :cursing_face: 
> Here is the red text for you:
>
>> System Message: Using QTextCodec: "**System**" from the available list:
"437",
"667",
"737",
"869",
...
"UTF-16",
"UTF-16BE",
"UTF-16LE",
"UTF-32",
"UTF-32BE",
"UTF-32LE",
"**UTF-8**",
"WINSAMI2",
...

We tell it to use a particular encoding, that encoding is clearly available - and it f\*\*\*\*\*g well ignores it!!!

Commit:
Testing: add an additional informational message about QTextCodec selection

This will report whether a codec was actually generated (as `pCodec` will
not be a `nullptr` in that case - and then report the name it has. This
should clarify whether one was correctly generated for UTF-8 as opposed to
the previous one that indicated that a local8Bit one was being used in some
cases on Windows (as it was called "system") instead of the expected UTF-8.

Signed-off-by: Stephen Lyons <stephen@Hunt>

Commit:
Bugfig: sort out ambiguous uses of QString::arg(...) with QByteArrays

Signed-off-by: Stephen Lyons <stephen@Hunt>

SlySven committed:
BugFix: try an alternative method to write config.lua file for Windows

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
BugFix: QDataStream does not have a flush() method

Unlike `QTextStream`...!

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
BugFix: correct methodology to get size of char array from a QByteArray

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Refactor: remove unneeded SLOT function

The same effect can just as easily (and faster) be done with a normal
method - and there is no need for an argument that is no-longer used since
the package exporter was redesigned by someone else after my previous work.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Cleanup: some ad-hoc modifications to the package exporter code

* adjust a replacement parameter in a `QObject::tr(...)` string
* relocate an explanation about not using the native file dialogue because
it has a particular limitation on Windoze to the point where the code it
applies to has been moved.
* fix some messed up indentation of a couple of multi-line comments
* unify some preprocessor includes to a `#if defined(...)` form

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Revise: make exporter tempory files be a subdirectory of a mudlet one

On some OSes the temporary files location in the filesystem is shared so
it is benefical for all the mudlet specific ones to be in a mudlet
subdirectory.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
BugFix: pass correct string length to zip_set_archive_comment(...)

The previous code was passing the length of a `QString` i.e. UTF-16BE
encoded data but the function needs to have the length of the string when
it is UTF-8 encoded (in bytes/chars) in a `QByteArray`. Whilst these values
will be the same if the string is purely an ASCII character one they will
be different as soon as the string contains characters outside of that
range of values.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven commented:
Humm, I've just read something that had been fouling my coding to try and set the Codec to use in the documentation for `QTextStream::setDevice(....)`:
> Sets the current device to device. If a device has already been assigned, QTextStream will call flush() before the old device is replaced.
>
> Note: This function resets locale to the default locale ('C') and codec to the default codec, QTextCodec::codecForLocale().

Guess who was calling `QTextStream::setCodec(...)` **before** `QTextStream::setDevice(...)` in *some* places...!

🤦‍♂️

vadi2 commented:
Good find!

SlySven committed:
Revise: ensure QTextCode::setCodec is used AFTER setDevice

I think this may have been the cause of the odd behaviour where the code
asked for the UTF-8 codec but ended up using a different one - the
setDevice(...) method happens to reset the codec so a setCodec(...) call
must be used after it and not before!

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

vadi2 commented:
There is a whole lot of unrelated changes again in e54bcef that have nothing to do with encoding issues - not OK, please do them separately.

SlySven commented:
Well TBH it *is* a W.I.P. - a draft - that I have been bashing away at for a couple of weeks when I found the time - it is just that I spotted **other** issues as well with the package exporter. If @Kebap reckons it does work I will try and distil out the other bits into other PRs that can be looked at separately.

:crossed_fingers: 

So far those other PRs are:
* #5356 
* #5357 
* #5358
* #5359
* #5360
* #5361
* #5363

SlySven committed:
Revert: remove other fixes (even if needed) so that PR stays on-topic

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Revert: missed from previous commit

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven commented on src/Host.cpp:
```diff
@@ -1973,9 +1980,17 @@ QString Host::getPackageConfig(const QString& luaConfig, bool isModule)
         while (!in.atEnd()) {
             strings += in.readLine();
         }
+        configFile.close();
```
Fair enough this isn't exactly on-topic but it is needed and since it is within 5 lines of a needed change it will be easier to include it I think rather than have to manually resolve a conflict between this and a separate PR to fix it.

Kebap commented:
> If @Kebap reckons it does work

the latest version did work!

SlySven commented:
>  the latest version did work!

Okay then - hopefully @vadi2 will approve it...

vadi2 commented on src/Host.cpp:
```diff
@@ -419,9 +419,16 @@ Host::Host(int port, const QString& hostname, const QString& login, const QStrin
     mErrorLogFile.open(QIODevice::Append);
-    // This is NOW used (for map
-    // file auditing and other issues)
+    // This is NOW used (for map file auditing and other issues)
```
Not so 'now' any more, this comment got stale. I don't think it adds anything anymore!

SlySven replied:
```suggestion
```

vadi2 commented on src/Host.cpp:
```diff
@@ -419,9 +419,16 @@ Host::Host(int port, const QString& hostname, const QString& login, const QStrin
-    // This is NOW used (for map
-    // file auditing and other issues)
+    // This is NOW used (for map file auditing and other issues)
+    /*
+     * Whilst we only generate ASCII characters in our messages that get logged
+     * there can easily be not ASCII in user content and if the logs get sent to
+     * someone else for inspection there is a fair chance that they do not have
+     * the same codec given that the default behaviour is to use
+     * QTextCodec::codecForLocale for QTextStream:
+     */
```
```suggestion
    // Mudlet will log messages in ASCII, but force UTF-8 in case user-generated content appears in the logs
```

SlySven replied:
```suggestion
     /*
     * Mudlet will log messages in ASCII, but force a universal (UTF-8) encoding
     * since user-content can contain anything and someone else reviewing
     * such logs need not have the same default encoding which would be used
     * otherwise - note that this must be done AFTER setDevice(...):
     */
```

vadi2 reviewed (commented):
Looks like a great fix, thanks for fighting Windows with it, that is never fun.

Brings us closer to taking off the experimental label off the package exporter, too!

vadi2 committed:
Update src/Host.cpp

Co-authored-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Update src/Host.cpp

Remove "NOW" comment that is not so up to the minute!
